config: Follow-ups to code resetting cpu/memory after preset change

cfergeau opened the pull request:
The main goal of this patch series is to fix 2 bugs related to https://github.com/crc-org/crc/pull/3636 :
- cpu/memory are not checked when the preset is unset, only when they are set 
- the messages instructing the user to restart the VM are no longer printed if cpu/memory are set to their default value

This starts by refactoring the code a bit, and by adding a new RegisterNotifier API, and then this fixes the bugs.
This also adds unit tests for this new code/these bug fixes.

**Fixes:** Issue #3652, Issue #3663

**Relates to:** Issue #3637

cfergeau committed:
config: Refactor code from commit 73a4b92d

The settings API already associates a settings name (CPUs, Memory) with
a validation function. This validation function takes into account
the current preset.

This commit reworks 73a4b92d to take advantage of this validation
function. Instead of reparsing the preset, and hardcoding in a new place
how to validate a Memory value, or a CPUs value, we can use instead the
validationFn from the Settings API.

cfergeau committed:
config: Refactor to avoid duplicated code

cfergeau committed:
config: Use notifier to update config on preset change

The RegisterNotifier API can be used to be notified when the preset
changes, and to check that the memory/cpu config values match the
requirements for the current preset.

The code in config.go is generic with respect to the config key names,
hardcoding in the generic Set() method that there will be settings named
CPUs/Preset/Memory is a layering violation. With a notifier, we can move
it back to settings.go which is where we define the settings (Preset,
CPUs, ...) we'll be using.

This will also fix https://github.com/crc-org/crc/issues/3652
as the notifiers are triggered both for Set() and Unset()

After this commit, this will work as expected:

crc config set preset podman
crc config set cpus 3
crc config unset preset
crc config get cpus

cfergeau committed:
config: Add RegisterNotifier

This allows to trigger a callback when a specific config key is modified
or unset, and to do some higher level actions in response to that.
This allows to remove the somewhat ugly UpdateDefaults() call from
the generic code in config.go, and to add this to RegisterSettings
instead where it makes more sense.

cfergeau committed:
version: Add GetDefaultPreset()

This will allow to override it at compile-time when needed.
I changed the default preset in the configuration code, but I did not
check if it was used in other places.

This will help with https://github.com/crc-org/crc/issues/3637

cfergeau committed:
config: Add unit test for set/unset preset

This checks that CPU/memory are updated as needed when moving from one
preset to another. If CPU/memory are invalid for the new preset, the
config code should reset them to their default value.

cfergeau committed:
config: use config notifier to call UpdateDefaults()

There are calls to config.UpdateDefaults in pkg/crc/api/ to reload the
configuration settings every time a config is set/unset. This ensures the
default cpus/memory values are correct when the preset changes.

Then there is code in pkg/crc/config.Set()/Unset() to reset the cpus/memory
values when the preset changes if they are out of spec.

This commit groups in UpdateDefaults all the work that is needed when
the preset changes. It moves the Set/Unset UpdateDefaults calls from
pkg/crc/api to pkg/crc/config

This should fix https://github.com/crc-org/crc/issues/3652

cfergeau committed:
config: set: Return callback after setting default value

This should fix https://github.com/crc-org/crc/issues/3663

Before this commit, the steps below

$ crc config set preset openshift
$ crc config set cpus 4

no longer print:

Changes to configuration property 'cpus' are only applied when the CRC instance is started.
If you already have a running CRC instance, then for this configuration change to take effect, stop the CRC instance with 'crc stop' and restart it with 'crc start'.

cfergeau commented on pkg/crc/config/settings_test.go:
```diff
@@ -0,0 +1,142 @@
+
+	_, err = cfg.Set(Preset, preset.Podman)
+	require.NoError(t, err)
+	_, err = cfg.Set(Memory, 10000)
```
CI/tests are failing because of this, when we set memory, the memory of the system is checked, and the runners only have 8GB. Hard to workaround in the tests as the minimum memory amount for OpenShift is 9GB. I guess I'll remove this part of the test, it's not the key part.
This will have to wait until next week though.

```
=== RUN   TestSetPreset
    settings_test.go:63: 
        	Error Trace:	/root/rpmbuild/BUILD/crc-2.19.0-4.12.13/pkg/crc/config/settings_test.go:63
        	Error:      	Received unexpected error:
        	            	Value '10000' for configuration property 'memory' is invalid, reason: only 7.281GB of memory found (10.49GB required)
        	Test:       	TestSetPreset
--- FAIL: TestSetPreset (0.00s)
```

praveenkumar replied:
Changing the cpu/memory to default we should use `podman` and `microshift` preset instead using `openshift` which fails because of high memory requirement.

gbraad replied:
I agree. Public CI's should only concern with the smaller presets we offer.

cfergeau replied:
Using microshift will paper over this problem for now, but it will break in the same way the day we get a worker with less than 4G, or the day we raise the microshift requirements.

cfergeau replied:
I started changing the test to use the microshift preset instead of openshift, but this does not allow me to test `crc config unset preset`, which is one of the bugs I'm fixing. I'll try to handle this differently.

gbraad commented:
I am confused about the naming `notifier` and the actual use. The commit says:

```
and to to some higher level actions in response to that.
```

so this is not to 'notify' the user, but an internal action. These concern ChangedValue functions, so can this be ChangeNotifier or similar... or as ChangedValueHandler? Also `notify` could be `notifyOfChange` or `handleValueChange`.

cfergeau commented:
> I am confused about the naming `notifier` and the actual use. The commit says:
> 
> ```
> and to to some higher level actions in response to that.
> ```
> 
> so this is not to 'notify' the user, but an internal action. These concern ChangedValue functions, so can this be ChangeNotifier or similar... or as ChangedValueHandler? Also `notify` could be `notifyOfChange` or `handleValueChange`.

I'll rename to ChangeNotifier.

cfergeau committed:
config: Don't export the validate* functions

They are not used outside of the `config` package except for ValidateBool
(which is used in the preflight package to register the skip-* settings,
and which is still exported after this commit)

Since validateMemory can be overridden, it's better if it's only
possible to do this from within the `config` package, ie it's better if
it's validateMemory() and not ValidateMemory().

anjannath commented on pkg/crc/config/settings.go:
```diff
@@ -130,6 +131,20 @@ func RegisterSettings(cfg *Config) {
 	cfg.AddSetting(IngressHTTPSPort, constants.OpenShiftIngressHTTPSPort, ValidatePort, RequiresHTTPSPortChangeWarning,
 		fmt.Sprintf("HTTPS port to use for OpenShift ingress/routes on the host (1024-65535, default: %d)", constants.OpenShiftIngressHTTPSPort))
+
+	_ = cfg.RegisterNotifier(Preset, presetChanged)
```
nit: did you consider logging the error in `cfg.RegisterNotifier()` instead of returning it, might lose some testability but we can get rid of the _ignore return value_

cfergeau replied:
hmm `config.go` does not log anything at the moment, I prefer to keep it this way (even though some low level debug logs there could be useful).
I'll log the error in `settings.go` though ;)

anjannath commented on pkg/crc/config/settings_test.go:
```diff
@@ -0,0 +1,163 @@
+	// Changing the preset to 'openshift' should reset the CPUs config value
+	logging.Warnf("Setting preset to OpenShift")
+	_, err = cfg.Set(Preset, crcpreset.OpenShift)
+	logging.Warnf("Preset set to OpenShift")
```
are these two `logging.Warnf` left over from testing?

cfergeau replied:
Yes, that's likely, I removed them, thanks!

anjannath commented on pkg/crc/config/viper_config_test.go:
```diff
@@ -327,3 +327,27 @@ func TestNotifier(t *testing.T) {
+	require.NoError(t, err)
+	require.NotEmpty(t, callback)
+
+	callback, err = config.Set(cpus, 4)
```
can also directly use the `constants.GetDefaultCPUs()` like the unit tests in the previous commit

cfergeau replied:
The rest of this file uses 4, 5 directly, I prefer to keep the 4 here as well.

anjannath commented on pkg/crc/config/settings.go:
```diff
@@ -46,16 +46,16 @@ func RegisterSettings(cfg *Config) {
 
-	validateCPUs := func(value interface{}) (bool, string) {
-		return ValidateCPUs(value, GetPreset(cfg))
+	validCPUs := func(value interface{}) (bool, string) {
```
should be `validateCPUs` ?

cfergeau replied:
this can't be `validateCPUs` as the name is already used by the function in validations.go. It's the same for `validMemory`/.. The `validCPUs` name is not great, open to suggestions! Maybe rename `validateCPUs` to `validateCPUsForPreset` ?

anjannath replied:
eh, no need, `validCPUs` is also okay :)

anjannath reviewed (commented):
like the addition of the change notifiers 👍🏽 just some small questions and comments.

also some of the commit msgs have small typos, listing the commits:

- 71c92b1cb7a016f7fa9d48af237d213a67df1c6f — s/is/in
- 574a883e240336518acf290ec1669a9ada12a130 — s/to/do
- 2120cdcb222755c7956dbbceb5c455f88ceca445 — s/save/except

cfergeau commented:
> also some of the commit msgs have small typos, listing the commits:
> [...]
>     * [2120cdc](https://github.com/crc-org/crc/commit/2120cdcb222755c7956dbbceb5c455f88ceca445) — s/save/except

"save for" is not a typo, see https://en.wiktionary.org/wiki/save_for
I've changed it to "except for" as you suggested, this will make the log easier to read

anjannath commented:
/retest

anjannath commented:
/test e2e-crc

anjannath commented:
proxy tests are failing in `e2e` are unrelated to the changes in the PR
